### PR TITLE
Adding intermediate challenges 

### PR DESCRIPTION
**Pull request by TsuyoshiUshio**
I add Serverless x DevOps challenges. 

@gled4er Could you review this, especially in terms of English? :)

### CODE REVIEW

File: 9-Serverless_Challenges/devopschallenges.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Serverless DevOps challenges
+
+This challenges help you to apply development environment and DevOps capability to your Serverless world.
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: I would say here: "This challenges help you to **configure** development environment and **add** DevOps capability to your Serverless world."

File: 9-Serverless_Challenges/devopschallenges.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# Serverless DevOps challenges
+
+This challenges help you to apply development environment and DevOps capability to your Serverless world. 
+
+
+## Challenge 1. Building your local environment
+
+As your success of the serverless architecture adoption, several people is joining your team.
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: Your project is getting very successful and popular, so new people are joining your team.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# Serverless DevOps challenges
+
+This challenges help you to apply development environment and DevOps capability to your Serverless world. 
+
+
+## Challenge 1. Building your local environment
+
+As your success of the serverless architecture adoption, several people is joining your team. 
+You need to have separate development environment with your colleague to develop in parallel.
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: You need to create separate development environments to develop in parallel with your colleagues.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+
+## Challenge 1. Building your local environment
+
+As your success of the serverless architecture adoption, several people is joining your team. 
+You need to have separate development environment with your colleague to develop in parallel. 
+
+### Requirement
+
+ You need to setup your local development environment on your PC. You can choose C# or Javascript
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: You need to setup your local development environment on your PC. You can choose to setup C# or Javascript based environment.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+As your success of the serverless architecture adoption, several people is joining your team. 
+You need to have separate development environment with your colleague to develop in parallel. 
+
+### Requirement
+
+ You need to setup your local development environment on your PC. You can choose C# or Javascript  
+
+### Success criteria
+
+* You can run your function locally using Storage Emulator
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: Shall we add full stop at the end of the sentences here?

File: 9-Serverless_Challenges/devopschallenges.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+* You can debug your function locally with a break point
+* You can push your function to the Azure
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: You can **deploy** your function to the Azure

File: 9-Serverless_Challenges/devopschallenges.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+### Reference
+
+* [Create your first function using Visual Studio](https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-your-first-function-visual-studio)
+* [Create your first function using Visual Studio Code](https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-first-function-vs-code)
+
+
+
+## Challenge 2. CI/ CD pipeline
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: **CI/CD** pipeline (without the space)

File: 9-Serverless_Challenges/devopschallenges.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+Your team need to share the code with your colleague to a private repository. Using Azure DevOps, set up the private repository with CI/CD pipeline
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: Your team needs to share the code **via** private repository. Using Azure DevOps, set up the private repository with CI/CD pipeline**.**

File: 9-Serverless_Challenges/devopschallenges.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+## Requirement
+
+Upload your code to the Azure DevOps with a very simple unit test code. Once you submit a pull request, it start build/test your pull request automatically. Also, once you check in to the master, it automatically build/test/deploy to Azure.
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: Upload your code to Azure DevOps with a very simple unit test**s**. Once you submit a pull request, it starts build/test **of the pull request's code** automatically. Also, once you **merge it** to the master **branch**, it automatically build/test/deploy to Azure.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+## Success criteria
+     
+* You can share the code in a private repo
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: Shall we add full stops at the end of each sentence?

Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: You can share the code via private repo.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+* Your pull request triggs a CI pipeline with at least one unit test code.
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: Your pull request **triggers** a CI pipeline with at least one unit test.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+* Automatically detect the change of master branch, triggs CI/CD pipeline to build/test/deploy to Azure.
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: Automatically detect**s** the change **in the** master branch **and** triggers CI/CD pipeline **for** build/test/deploy to Azure.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+### Reference 
+
+* [Azure Repos Git Tutorial](https://docs.microsoft.com/en-us/azure/devops/repos/git/gitworkflow?view=vsts)
+* [Serverless, DevOps, and CI/CD: Part 2](https://medium.com/microsoftazure/serverless-devops-and-ci-cd-part-2-b6e0a6d05530)
+* [Azure Functions CI/CD pipeline for Node.js using VSTS](https://blogs.technet.microsoft.com/livedevopsinjapan/2017/12/13/azure-functions-cicd-pipeline-for-node-js-using-vsts/)
+
+## Challenge 3. Run on container (Preview)
+
+As your function is totally available, people from other division want to use it. However, they don't have Azure environment. They required it works on premise environment. You need to run your function locally on container.
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: However, they don't have access to an Azure environment. They want to run your function on premise environment. You need to package your function in a Docker container.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+* [Serverless, DevOps, and CI/CD: Part 2](https://medium.com/microsoftazure/serverless-devops-and-ci-cd-part-2-b6e0a6d05530)
+* [Azure Functions CI/CD pipeline for Node.js using VSTS](https://blogs.technet.microsoft.com/livedevopsinjapan/2017/12/13/azure-functions-cicd-pipeline-for-node-js-using-vsts/)
+
+## Challenge 3. Run on container (Preview)
+
+As your function is totally available, people from other division want to use it. However, they don't have Azure environment. They required it works on premise environment. You need to run your function locally on container. 
+
+## Requirement
+
+Pack your function into docker container and run on your local machine with sharing on DockerHub or Azure Container Registry then run on Azure
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: Package your function into a Docker container and run it on your local machine. Then share the container via DockerHub or Azure Container Registry. The final step is to run it on Azure.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+## Success criteria
+
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: Shall we add full stops at the end of each sentence?

File: 9-Serverless_Challenges/devopschallenges.md [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+Your container works on your local machine as a docker container
+Your image is uploaded to the DockerHub or Azure Container Registry
+Run the docker container on Azure
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: You are able to run the Docker container on Azure.

File: 9-Serverless_Challenges/devopschallenges.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+Pack your function into docker container and run on your local machine with sharing on DockerHub or Azure Container Registry then run on Azure
+
+## Success criteria
+
+Your container works on your local machine as a docker container
+Your image is uploaded to the DockerHub or Azure Container Registry
+Run the docker container on Azure 
+
+NOTE: You have three ways to run this. Azure Functions, Azure Container Instances, Azure Kubernetes Services
```
Comment (gled4er) [inline 9-Serverless_Challenges/devopschallenges.md]: You have three ways to run the Docker container in Azure:

1. Azure Functions
2. Azure Container Instances
3. Azure Kubernetes Services

**Review by gled4er — CHANGES_REQUESTED** [review]

File: 9-Serverless_Challenges/devopschallenges.md (hunk 1 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done.

File: 9-Serverless_Challenges/devopschallenges.md (hunk 2 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 3 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 4 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 6 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 5 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 7 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 8 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 9 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 10 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 11 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 12 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 13 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 14 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 15 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done.

File: 9-Serverless_Challenges/devopschallenges.md (hunk 16 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

File: 9-Serverless_Challenges/devopschallenges.md (hunk 17 above)
Comment (TsuyoshiUshio) [reply 9-Serverless_Challenges/devopschallenges.md]: Done

**Comment by TsuyoshiUshio** [comment]
Hi @gled4er I reflect all changes which you recommended. :)

**Review by gled4er — APPROVED** [review]
Great job on these challenges!

**Pull request merged**